// @pixi/core/lib/index.mjs
import './settings.mjs';
export * from '@pixi/color';
export * from '@pixi/constants';
export * from '@pixi/extensions';
export * from '@pixi/math';
export * from '@pixi/runner';
export * from '@pixi/settings';
export * from '@pixi/ticker';
import * as utils$1 from '@pixi/utils';
export { utils$1 as utils };
export { autoDetectRenderer } from './autoDetectRenderer.mjs';
export { BackgroundSystem } from './background/BackgroundSystem.mjs';
export { BatchDrawCall } from './batch/BatchDrawCall.mjs';
export { BatchGeometry } from './batch/BatchGeometry.mjs';
export { BatchRenderer } from './batch/BatchRenderer.mjs';
export { BatchShaderGenerator } from './batch/BatchShaderGenerator.mjs';
export { BatchSystem } from './batch/BatchSystem.mjs';
export { BatchTextureArray } from './batch/BatchTextureArray.mjs';
export { ObjectRenderer } from './batch/ObjectRenderer.mjs';
export { ContextSystem } from './context/ContextSystem.mjs';
export { Filter } from './filters/Filter.mjs';
export { FilterState } from './filters/FilterState.mjs';
export { FilterSystem } from './filters/FilterSystem.mjs';
import './filters/IFilterTarget.mjs';
export { SpriteMaskFilter } from './filters/spriteMask/SpriteMaskFilter.mjs';
export { defaultFilterVertex, defaultVertex } from './fragments/index.mjs';
export { Framebuffer } from './framebuffer/Framebuffer.mjs';
export { FramebufferSystem } from './framebuffer/FramebufferSystem.mjs';
export { GLFramebuffer } from './framebuffer/GLFramebuffer.mjs';
export { MultisampleSystem } from './framebuffer/MultisampleSystem.mjs';
export { Attribute } from './geometry/Attribute.mjs';
export { Buffer } from './geometry/Buffer.mjs';
export { BufferSystem } from './geometry/BufferSystem.mjs';
export { Geometry } from './geometry/Geometry.mjs';
export { GeometrySystem } from './geometry/GeometrySystem.mjs';
export { ViewableBuffer } from './geometry/ViewableBuffer.mjs';
import './IRenderer.mjs';
export { MaskData } from './mask/MaskData.mjs';
export { MaskSystem } from './mask/MaskSystem.mjs';
export { ScissorSystem } from './mask/ScissorSystem.mjs';
export { StencilSystem } from './mask/StencilSystem.mjs';
export { PluginSystem } from './plugin/PluginSystem.mjs';
export { ProjectionSystem } from './projection/ProjectionSystem.mjs';
export { ObjectRendererSystem } from './render/ObjectRendererSystem.mjs';
export { Renderer } from './Renderer.mjs';
export { BaseRenderTexture } from './renderTexture/BaseRenderTexture.mjs';
export { GenerateTextureSystem } from './renderTexture/GenerateTextureSystem.mjs';
export { RenderTexture } from './renderTexture/RenderTexture.mjs';
export { RenderTexturePool } from './renderTexture/RenderTexturePool.mjs';
export { RenderTextureSystem } from './renderTexture/RenderTextureSystem.mjs';
export { GLProgram, IGLUniformData } from './shader/GLProgram.mjs';
export { Program } from './shader/Program.mjs';
export { Shader } from './shader/Shader.mjs';
export { ShaderSystem } from './shader/ShaderSystem.mjs';
export { UniformGroup } from './shader/UniformGroup.mjs';
export { checkMaxIfStatementsInShader } from './shader/utils/checkMaxIfStatementsInShader.mjs';
export { generateProgram } from './shader/utils/generateProgram.mjs';
export { createUBOElements, generateUniformBufferSync, getUBOData } from './shader/utils/generateUniformBufferSync.mjs';
export { getTestContext } from './shader/utils/getTestContext.mjs';
export { uniformParsers } from './shader/utils/uniformParsers.mjs';
export { unsafeEvalSupported } from './shader/utils/unsafeEvalSupported.mjs';
export { StartupSystem } from './startup/StartupSystem.mjs';
export { State } from './state/State.mjs';
export { StateSystem } from './state/StateSystem.mjs';
import './system/ISystem.mjs';
import './systems.mjs';
export { BaseTexture } from './textures/BaseTexture.mjs';
export { GLTexture } from './textures/GLTexture.mjs';
import './textures/resources/index.mjs';
export { Texture } from './textures/Texture.mjs';
export { TextureGCSystem } from './textures/TextureGCSystem.mjs';
export { TextureMatrix } from './textures/TextureMatrix.mjs';
export { TextureSystem } from './textures/TextureSystem.mjs';
export { TextureUvs } from './textures/TextureUvs.mjs';
export { TransformFeedback } from './transformFeedback/TransformFeedback.mjs';
export { TransformFeedbackSystem } from './transformFeedback/TransformFeedbackSystem.mjs';
export { Quad } from './utils/Quad.mjs';
export { QuadUv } from './utils/QuadUv.mjs';
export { ViewSystem } from './view/ViewSystem.mjs';
export { SystemManager } from './system/SystemManager.mjs';
export { BaseImageResource } from './textures/resources/BaseImageResource.mjs';
export { Resource } from './textures/resources/Resource.mjs';
export { AbstractMultiResource } from './textures/resources/AbstractMultiResource.mjs';
export { ArrayResource } from './textures/resources/ArrayResource.mjs';
export { INSTALLED, autoDetectResource } from './textures/resources/autoDetectResource.mjs';
export { BufferResource } from './textures/resources/BufferResource.mjs';
export { CanvasResource } from './textures/resources/CanvasResource.mjs';
export { CubeResource } from './textures/resources/CubeResource.mjs';
export { ImageBitmapResource } from './textures/resources/ImageBitmapResource.mjs';
export { ImageResource } from './textures/resources/ImageResource.mjs';
export { SVGResource } from './textures/resources/SVGResource.mjs';
export { VideoResource } from './textures/resources/VideoResource.mjs';

const VERSION = "7.2.4";

export { VERSION };
